src/hooks/browser-headers.ts

```diff
@@ -22,40 +22,69 @@ export function mergeHeaders(original: Record<string, string>, overrides: Record
     return Object.fromEntries(fixedHeaders.values());
 }
 
+interface StoredHeaders {
+    1: Record<string, string>;
+    2: Record<string, string>;
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+const headers = new WeakMap<object, StoredHeaders>();
+
 export async function browserHeadersHook(options: Options): Promise<void> {
     const { context } = options;
     const {
         headerGeneratorOptions,
         useHeaderGenerator,
         headerGenerator,
         proxyUrl,
+        sessionToken,
     } = context as Context;
 
     if (!useHeaderGenerator || !headerGenerator) return;
 
     const url = options.url as URL;
 
-    const resolveProtocol = proxyUrl ? createResolveProtocol(proxyUrl) : http2.auto.resolveProtocol;
+    const resolveProtocol = (options as any).resolveProtocol
+        || (proxyUrl
+            ? createResolveProtocol(proxyUrl)
+            : http2.auto.resolveProtocol);
 
     let alpnProtocol;
     if (url.protocol === 'https:') {
         alpnProtocol = (await resolveProtocol({
             host: url.hostname,
             port: url.port || 443,
             rejectUnauthorized: false,
-            // @ts-expect-error Open an issue in http2-wrapper
             ALPNProtocols: ['h2', 'http/1.1'],
             servername: url.hostname,
         })).alpnProtocol;
     }
 
-    const mergedHeaderGeneratorOptions = {
-        httpVersion: alpnProtocol === 'h2' ? '2' : '1',
-        ...headerGeneratorOptions,
-    };
+    const httpVersion = alpnProtocol === 'h2' ? '2' : '1';
 
-    const generatedHeaders = headerGenerator.getHeaders(mergedHeaderGeneratorOptions);
+    let generatedHeaders: Record<string, string>;
+    if (sessionToken) {
+        if (!headers.has(sessionToken)) {
+            headers.set(sessionToken, {
+                1: headerGenerator.getHeaders({
+                    httpVersion: '1',
+                    ...headerGeneratorOptions,
+                }),
+                2: headerGenerator.getHeaders({
+                    httpVersion: '2',
+                    ...headerGeneratorOptions,
+                }),
+            });
+        }
+
+        generatedHeaders = headers.get(sessionToken)![httpVersion];
+    } else {
+        generatedHeaders = headerGenerator.getHeaders({
+            httpVersion,
+            ...headerGeneratorOptions,
+        });
+    }
 
-    // TODO: Remove this when Got supports Headers class.
+    // TODO: Use `options.merge({headers: generatedHeaders})` instead
```

Comment:
Use when?

Reply:
When the `options` test object is an actual `Options` instance :) This can be refactored later.